### ISSUE TITLE
Model converted but can't run

### ISSUE THREAD

**Issue by gasgallo** [XiaoMi/mace#341]
I'm trying to test a tensorflow model on a Xiaomi smartphone.

Model conversion is successful, but the inference isn't. Reported codes are:

`F eltwise.h:80 Check failed: input0->dim(3) == input1->dim(0) Element-Wise op support broadcast on only-channel or non-channel dimension`

and

`F eltwise.cc:858 Check failed: input0->dim(rank_diff + i) == 1 || input1->dim(i) == 1 || input0->dim(rank_diff + i) == input1->dim(i) Element-Wise op only support tail dimensions broadcast`

Is some operation in my model not supported by MACE? How can I identify the op source of the problem?

Thanks.

**Comment by llhe** [comment]
For the 1st issue, the problem is caused by this check:
https://github.com/XiaoMi/mace/blob/master/mace/ops/opencl/image/eltwise.h#L80

Which means MACE (OpenCL+GPU mode) does not currently support  such elementwise op with flexible dim broadcast. Could you try to remove the broadcast (see https://databricks.com/tensorflow/dimensionality-and-broadcasting).

**Comment by gasgallo** [comment]
@llhe 

Thanks for the hint! I've tried to specify all the dimensions of the tensors but I get back the same errors.
Is there a way to identify the specific op that causes the problem?

**Comment by llhe** [comment]
@gasgallo You can try to debug it with more logs to see which op failed: https://mace.readthedocs.io/en/latest/development/how_to_debug.html#debug-engine-using-log

**Comment by gasgallo** [comment]
@llhe 

This is the last part of the log:

'I logging.h:105 Initializing SerialNet started
I logging.h:111 Initializing SerialNet latency: 1040 us
I mace_run.cc:282 Total init latency: 103.632 ms
I mace_run.cc:322 Warm up run
I mace.cc:545 Transform input data from NHWC to NCHW
I logging.h:105 Running net started
I logging.h:105 Running operator Sub<0, Eltwise, 1> started
F eltwise.cc:858 Check failed: input0->dim(rank_diff + i) == 1 || input1->dim(i) == 1 || input0->dim(rank_diff + i) == input1->dim(i) Element-Wise op only support tail dimensions broadcast
Aborted
ERROR: [Mace Run] Mace run failed.'

It's not straightforward to interpret the meaning of 'Sub<0, Eltwise, 1> ' because there's no direct reference to any of the op of my network.
How can I better interpret the log?

**Comment by llhe** [comment]
@gasgallo Assume you are using TensorFlow, you can use tensorboard to visualize your graph to identify this Op.

**Comment by gasgallo** [comment]
@llhe 

All right, I got to identify the Op, but I don't understand why some broadcast takes place given that the inputs have the same dimensions.

![screenshot from 2019-02-14 09-56-22](https://user-images.githubusercontent.com/36033044/52759611-15e65400-303f-11e9-9d77-7a7a382905c5.png)

**Comment by nolanliou** [comment]
@gasgallo you could print the inputs' shape of the sub(`Eltwise`) op to check the correctness.

**Comment by gasgallo** [comment]
@nolanliou you were right!!

`input0 1, 3, 112, 112`
`input1 1, 112, 112, 3`

For some reason it messed the tensor shape. How to solve?

**Comment by nolanliou** [comment]
The `input1` is const tensor?

**Comment by gasgallo** [comment]
> The `input1` is const tensor?

Yes

**Comment by nolanliou** [comment]
So, that's the point. We have not encountered this usage yet, so MACE does not support the case. Generally, the inputs generated by other operation if it's shape is 4D.

**Comment by gasgallo** [comment]
> So, that's the point. We have not encountered this usage yet, so MACE does not support the case. Generally, the inputs of `Eltwise` are generated by other operations if it's shape is 4D.

So, what if I specify the constant as a scalar?

**Comment by nolanliou** [comment]
Scalar and vector are supported.

**Comment by gasgallo** [comment]
@nolanliou 

If I use a tf.constant I get the following error:

`I logging.h:105 Initializing SerialNet started`
`I logging.h:111 Initializing SerialNet latency: 1015 us`
`I mace_run.cc:282 Total init latency: 106.58 ms`
`I mace_run.cc:322 Warm up run`
`I mace.cc:545 Transform input data from NHWC to NCHW`
`I logging.h:105 Running net started`
`I logging.h:105 Running operator Sub<0, Eltwise, 1> started`
`F tensor.h:205 Check failed: index < shape_.size() minusscalar0_second:0: Dim out of range: 1 >= 1`
`Aborted`
`ERROR: [Mace Run] Mace run failed.`

**Comment by nolanliou** [comment]
I have tested the below simple model and it's ok. Is there some difference about your code?
```
  conv = tf.layers.conv2d(                                                      
      inputs=input_layer,                                                       
      filters=3,                                                                
      kernel_size=[3, 3],                                                       
      padding="same",                                                           
      activation=tf.nn.relu6,                                                   
      name="conv")                                                              
  b = tf.constant(1, dtype=tf.float32)                                          
  res = tf.subtract(conv, b, name=output_node_name)                             
  return res 
```

**Comment by gasgallo** [comment]
@nolanliou 

I was passing the scalar as a ndarray, passing it as a float number solves the problem.
Now it gets stuck right when computing the output node with the following error:

`I logging.h:111 Running net latency: 581505 us`
`F mace.cc:636 Check failed: shape == output->second.shape() Output shape mismatch: [1, 512, 1, 1] != [1, 512]`
`Aborted`
`ERROR: [Mace Run] Mace run failed.`

I expect my output to be as the `output` tensor ([1, 512]), but seems like MACE expects a 4-dim tensor

**Comment by nolanliou** [comment]
You should keep the output shape consistent betwwen model deployment file(.yml) and c++ code.

**Comment by gasgallo** [comment]
Awesome, it's working now!
Thank you @nolanliou @llhe

**Comment by gasgallo** [comment]
I'm back with the same problem when using a vector instead of a scalar.

I've tried representing the vector as np.array and as list but both fails with the same error:
```
F mace/ops/eltwise.cc:866] Check failed: input0->dim(rank_diff + i) == 1 || input1->dim(i) == 1 || input0->dim(rank_diff + i) == input1->dim(i) Element-Wise op only support tail dimensions broadcast
```

The op is a Multiply and dimensions of operands are: 1,64,112,112 and 64 (debugged from eltwise.cc)

@nolanliou what am I doing wrong?

**Comment by gasgallo** [comment]
I've done some further testing and worked around the problem using a different Op.
I used `tf.nn.batch_normalization` with the following params:
- input: the first operand of tf.multiply
- mean: tf.zeros
- variance: tf.ones
- offset: None
- scale: the second operand of tf.multiply (as a vector, exactly the same that throws the error)
- variance_epsilon: 0

Basically I've used a batch normalization to run a multiplication.

Why does this work and tf.multiply doesn't?
